spec/integration/websocket_spec.lua

```diff
@@ -0,0 +1,100 @@
+    url = url or local_websocket_url
+    properties = properties or {}
+    local connection_properties = ConnectionProperties:create(properties)
+    local sock = websocket.connect(url, connection_properties)
+    finally(function() sock:close() end)
+    return sock
+end
+
+local function assert_connect_successful(properties)
+    local sock = connect(properties)
+    assert.is_not_nil(sock)
+    assert.is_false(sock.closed)
+    local result = sock:send_raw('{"command": "login", "protocolVersion": 3}')
+    assert.matches('.*"status":"ok".*', result)
+end
+
+local function assert_connect_fails(properties, expected_error)
+    expected_error = expected_error or ssl_negotion_failed_error
+    assert.error(function() connect(properties) end, expected_error)
+end
+
+describe("Websocket", function()
+    it("connects with default properties", function() --
+        assert_connect_successful()
+    end)
+
+    it("connects with tls_verify=none", function() --
+        assert_connect_successful({tls_verify = "none"})
+    end)
+
+    describe("fails connection with tls_verify option", function()
+        for _, tls_verify_option in ipairs({"invalid", "", "peer", "client_once", "fail_if_no_peer_cert"}) do
+            it(string.format("value %q", tls_verify_option), function() --
+                assert_connect_fails({tls_verify = tls_verify_option})
+            end)
+        end
+    end)
+
+    describe("connects with tls_protocol option", function()
+        for _, tls_protocol_option in ipairs({"any", "tlsv1_2"}) do
+            it(string.format("value %q", tls_protocol_option), function() --
+                assert_connect_successful({tls_protocol = tls_protocol_option})
+            end)
+        end
+    end)
+
+    describe("fails connection with tls_protocol option", function()
+        for _, tls_protocol_option in ipairs({"invalid", "", "tlsv1", "tlsv1_1", "tlsv1_3"}) do
+            it(string.format("value %q", tls_protocol_option), function() --
+                assert_connect_fails({tls_protocol = tls_protocol_option})
+            end)
+        end
+    end)
+
+    describe("connects with tls_options option", function()
+        for _, tls_options in ipairs({
+            "", "all", "no_tlsv1_1", "no_tlsv1_3", "no_tlsv1", "no_tlsv1_1,no_tlsv1_3,no_tlsv1"
+        }) do
+            it(string.format("value %q", tls_options),
+               function() assert_connect_successful({tls_options = tls_options}) end)
+        end
+    end)
+
+    describe("fails connection with tls_options option", function()
+        for _, tls_options in ipairs({
+            "no_tlsv1_2", "no_tlsv1_2,all", "no_tlsv1_2, all", "no_tlsv1_1,no_tlsv1_3, no_tlsv1"
+        }) do
+            it(string.format("value %q", tls_options), function()
+                assert_connect_fails({tls_options = tls_options})
+            end)
+        end
+    end)
+
+    it("log available ssl configuration options, useful for debugging", function()
```

Comment:
This is not really a test, since you are not exercising any production code. How about packing that into the "after all" part and simply outputting it there? That way the number of test cases still reflects actual tests.

Reply:
Good idea!